lib/srv/ctx.go

```diff
@@ -584,6 +603,136 @@ func (c *ServerContext) getSession() *session {
 	return c.session
 }
 
+// OpenXServerListener opens a new XServer unix listener.
+func (c *ServerContext) OpenXServerListener(x11Req x11.ForwardRequestPayload, displayOffset, maxDisplays int) error {
+	l, display, err := x11.OpenNewXServerListener(displayOffset, maxDisplays, x11Req.ScreenNumber)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.setX11Config(&X11Config{
+		XServerUnixSocket: l.Addr().String(),
+		XAuthEntry: &x11.XAuthEntry{
+			Display: display,
+			Proto:   x11Req.AuthProtocol,
+			Cookie:  x11Req.AuthCookie,
+		},
+	})
+	if err != nil {
+		l.Close()
+		return trace.Wrap(err)
+	}
+
+	c.AddCloser(l)
+
+	// Prepare X11 channel request payload
+	originHost, originPort, err := net.SplitHostPort(c.ServerConn.LocalAddr().String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	originPortI, err := strconv.Atoi(originPort)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	x11ChannelReqPayload := ssh.Marshal(x11.ChannelRequestPayload{
+		OriginatorAddress: originHost,
+		OriginatorPort:    uint32(originPortI),
+	})
+
+	go func() {
+		for {
+			xconn, err := l.Accept()
+			if err != nil {
+				// listener is closed
+				return
+			}
+
+			go func() {
+				defer xconn.Close()
+
+				// If the session has not signaled that X11 forwarding is
+				// fully set up yet, then ignore any incoming connections.
+				// The client's session hasn't been fully set up yet so this
+				// could potentially be a break-in attempt.
+				if ok, err := c.x11Ready(); err != nil {
+					log.WithError(err).Debug("Failed to get X11 ready status")
+					return
+				} else if !ok {
+					log.WithError(err).Debug("Rejecting X11 request, XServer Proxy is not ready")
+					return
+				}
+
+				xchan, sin, err := c.ServerConn.OpenChannel(sshutils.X11ChannelRequest, x11ChannelReqPayload)
+				if err != nil {
+					log.WithError(err).Debug("Failed to open a new X11 channel")
+					return
+				}
+				defer xchan.Close()
+
+				// Forward ssh requests on the X11 channels until X11 forwarding is complete
+				ctx, cancel := context.WithCancel(c.cancelContext)
+				defer cancel()
+
+				go func() {
+					err := sshutils.ForwardRequests(ctx, sin, c.RemoteSession)
+					if err != nil {
+						log.WithError(err).Debug("Failed to forward ssh request from server during X11 forwarding")
+					}
+				}()
+
+				if err := x11.Forward(ctx, xconn, xchan); err != nil {
+					log.WithError(err).Debug("Encountered error during X11 forwarding")
+				}
+			}()
+
+			if x11Req.SingleConnection {
+				l.Close()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// getX11Config gets the x11 config for this server session.
+func (c *ServerContext) getX11Config() *X11Config {
```

Comment:
Would it be better to return a copy `X11Config` instead of a pointer?

I'm wondering the caller can mutate the config without locking and introduce a race condition.

Reply:
I've updated X11Config to contain no pointers and am returning a shallow copy here instead, is that fine?

Reply:
Should be.